test/ViewExtensionLibraryTests/LibraryResourceProviderTests.cs

```diff
@@ -290,13 +293,212 @@ public void PackagedCustomNodeSearchElementLoadedType()
             var item = provider.CreateLoadedTypeItem(element);
             Assert.AreEqual(expectedQualifiedName, item.fullyQualifiedName);
             Assert.AreEqual(guid.ToString(), item.contextData);
             Assert.AreEqual(string.Empty, item.keywords);
             var url = new IconUrl(name, path, true);
             Assert.AreEqual(url.Url, item.iconUrl);
         }
+
+        [Test]
+        [Category("UnitTests")]
+        public void LibraryDataUpdatedEventRaised()
+        {
+            const string libraryDataUpdated = "libraryDataUpdated";
+            var timeout = 50; //50 milliseconds
+            var resetevent = new AutoResetEvent(false);
+
+            var model = new NodeSearchModel();
+            var controller = new Mock<IEventController>();
+            controller.Setup(c => c.RaiseEvent(It.IsAny<string>(), It.IsAny<object[]>())).Callback(() => resetevent.Set());
+
+            var disposable = LibraryViewController.SetupSearchModelEventsObserver(model, controller.Object, timeout);
+            controller.Verify(c => c.RaiseEvent(libraryDataUpdated, It.IsAny<object[]>()), Times.Never);
+
+            var d1 = MockNodeSearchElement("A", "B");
+            var d2 = MockNodeSearchElement("C", "D");
+            var d3 = MockNodeSearchElement("E", "F");
+            model.Add(d1.Object);
+            model.Add(d2.Object);
+            model.Add(d3.Object);
+            Assert.AreEqual(3, model.NumElements);
+
+            Assert.IsTrue(resetevent.WaitOne(timeout*3));
+            controller.Verify(c => c.RaiseEvent("libraryDataUpdated", "A, C, E"), Times.Once);
+
+            //Dispose
+            disposable.Dispose();
+            d1 = MockNodeSearchElement("G", "B");
+            d2 = MockNodeSearchElement("H", "D");
+            d3 = MockNodeSearchElement("I", "F");
+            model.Add(d1.Object);
+            model.Add(d2.Object);
+            model.Add(d3.Object);
+            Assert.AreEqual(6, model.NumElements);
+            controller.Verify(c => c.RaiseEvent(libraryDataUpdated, It.IsAny<object[]>()), Times.Once);
+        }
+
+        [Test]
+        [Category("UnitTests")]
+        public void SimpleEventObserver()
+        {
+            var controller = new Mock<IEventController>();
+            var observer = new EventObserver<int, bool>(
+                    x => controller.Object.RaiseEvent("X", x),
+                    x => x%2 == 0
+                );
+
+            var list = Enumerable.Range(1, 10).ToList();
+            list.ForEach(x => observer.OnEvent(x)); //notify OnEvent
+            controller.Verify(c => c.RaiseEvent("X", It.IsAny<bool>()), Times.Exactly(10));
+            controller.Verify(c => c.RaiseEvent("X", true), Times.Exactly(5));
+            controller.Verify(c => c.RaiseEvent("X", false), Times.Exactly(5));
+        }
+
+        [Test]
+        [Category("UnitTests")]
+        public void SimpleEvenNumberObserver()
```

Comment:
Yes, but the throttle is an additional operator, it just observes for even numbers only. Maybe it could be just `EvenNumberObserver`.